src/server/session.ts

```diff
@@ -691,13 +690,30 @@ namespace ts.server {
                 };
             }
 
-            return definitionAndBoundSpan;
+            return {
+                ...definitionAndBoundSpan,
+                definitions: definitionAndBoundSpan.definitions.map(Session.mapToOriginalLocation)
+            };
         }
 
         private mapDefinitionInfo(definitions: ReadonlyArray<DefinitionInfo>, project: Project): ReadonlyArray<protocol.FileSpan> {
             return definitions.map(def => this.toFileSpan(def.fileName, def.textSpan, project));
         }
 
+        private static mapToOriginalLocation<T extends DocumentSpan>(def: T): T {
+            if (def.originalFileName) {
+                Debug.assert(def.originalTextSpan !== undefined, "originalTextSpan should be present if originalFileName is");
+                return {
+                    ...<any>def,

```

Comment:
Lemme guess: excess property warning?

Reply:
"Spread types may only be created from object types" (even if `T` is constrained to `DocumentSpan & Object`!)

Reply:
This also requires `any` rather than a different cast because we don't believe that the returned object literal actually conforms to `T`

Reply:
So it's a symptom of how we don't have higher order spread types. :3